### PR TITLE
WIP Tcp conntrack

### LINKED ISSUE #101607: TCP Conntrack entries remain in stale
We have a service with type as LoadBalancer. The service external-IP is allocated by a controller and the this IP is known in advance even before the service is deployed. The service external IP is configured on the client machines located outside the Kubernetes cluster and they are trying to reach the service all the time. The adjacent routers are configured to forward the traffic matching the service external-IP towards the worker node statically.

In the beginning the clients are trying to reach to the service external-IP and at this time the Kubernetes cluster do not have the service or the pods up and running. At this time the iptables do not have entry for the service external-IP and this means that the packets will not be handled by the iptables populated by kube-proxy. These TCP SYN packets creates a conntrack entry which will get removed in 120 seconds if no further packets are matching to the tuple

sudo conntrack -L  | grep 30.4.5.22
conntrack v1.4.4 (conntrack-tools): 48 flow entries have been shown.
tcp      6 48 SYN_SENT src=137.64.100.78 dst=30.4.5.22 sport=63302 dport=80 [UNREPLIED] src=30.4.5.22 dst=137.64.100.78 sport=80 dport=63302 mark=0 use=1

In this case the service external IP is 30.4.5.22

The client is having limited (may be 3) TCP source ports due to the way firewalling and clients are configured. This means that clients keep trying to establish TCP session with same source port and causes the conntrack entry to be alive and not cleaned up at all.

When the service and pods are deployed at a later stage, the iptables are correctly updated with the service externalIP (30.4.5.22) rules. Unfortunately due to existing conntrack entry matching the packets destined to 30.4.5.22, the packets do not hit the NAT table. This means that the clients can not talk to the real service and they do not recover. If the conntrack entries are dropped after NAT tables are updated then the TCP sessions are established correctly.

This issue can happen in some cases of node reboots since there are cases where the external traffic arrives to the node before the kube-proxy can setup the NAT entries for the service external IP.

When we checked the kube proxy code we see that there are logic to remove the conntrack entries matching the service IP in case of UDP and SCTP but not for TCP.

Does it make the system robust if we clear the conntrack entry matching the IP that is added to the NAT table so that we do not have undesired behaviour due to stale conntrack entries?

#### Environment:
Kubernetes version is v1.20.4
kube-proxy mode is iptables

### PR DESCRIPTION
**Pull request by aojea**
/kind bug

kube-proxy was always deleting stale conntrack entries for stateless protocols, like UDP and SCTP, since TCP clients retries with different source ports and eventually the stale entries timeout.
However, on TCP clients with a low number of source ports, the stale entries can also blackhole connections, so we need to delete them too.

However, we should be careful deleting conntrack entries, since it can break TCP graceful shutdown, i.e. pods are still using the connection despite the endpoint is no longer present in kube-proxy. To avoid that, we delete only conntrack entries with the SYN_SENT status.

Fixes #101607

```release-note
kube-proxy was deleting stale conntrack entries for stateless protocols, like UDP and SCTP. However, on TCP clients with a low number of source ports, the stale entries can also blackhole connections, so we delete them too.
```

### CODE REVIEW

**Comment by aojea** [comment]
/hold
Still wip, I need to add more testing

File: test/e2e/network/conntrack.go [hunk 1]
```diff
@@ -279,6 +279,158 @@ var _ = common.SIGDescribe("Conntrack", func() {
...
+		fr.PodClient().CreateSync(clientPod)
+
+		// Read the client pod logs
+		logs, err := e2epod.GetPodLogs(cs, ns, podClient, podClient)
+		framework.ExpectNoError(err)
+		framework.Logf("Pod client logs: %s", logs)
+
+		// Add a backend pod to the service in the other node
+		ginkgo.By("creating a backend pod " + podBackend1 + " for the service " + serviceName)
+		serverPod1 := e2epod.NewAgnhostPod(ns, podBackend1, nil, nil, nil, "netexec", fmt.Sprintf("--tcp-port=%d", 80))
```
Comment (aojea) [inline test/e2e/network/conntrack.go:312]: this doesn't exist :upside_down_face: 
Thanks @hanamantagoudvk for noticing it

File: test/e2e/network/conntrack.go [hunk 2]
```diff
@@ -279,6 +279,158 @@ var _ = common.SIGDescribe("Conntrack", func() {
...
+			}
+		})
+		framework.ExpectNoError(err)
+
+		// Create a pod in one node to create the UDP traffic against the NodePort service every 5 seconds
+		ginkgo.By("creating a client pod for probing the service " + serviceName)
+		clientPod := e2epod.NewAgnhostPod(ns, podClient, nil, nil, nil)
+		nodeSelection := e2epod.NodeSelection{Name: clientNodeInfo.name}
+		e2epod.SetNodeSelection(&clientPod.Spec, nodeSelection)
+		cmd := fmt.Sprintf(`date; for i in $(seq 1 3000); do echo "$(date) Try: ${i}"; echo hostname | nc -w 5 -p %d %s %d; echo; done`, srcPort, serverNodeInfo.nodeIP, tcpService.Spec.Ports[0].NodePort)
```
Comment (hanamantagoudvk) [inline test/e2e/network/conntrack.go:300]: Even with netcat command , i am not able to see TCP SYN packet going. 
I tried this below option manually ,
nc -w 5 -p <src-port> <cluster-ip> <target-port>

**Comment by aojea** [comment]
let's move this problem to the application, TCP already has the tools and features needed to solve the conntrack problems

**Pull request closed without merging**